test/integration/controller_test.go

```diff
@@ -1319,3 +512,352 @@ func getTestHappyPathBrokerClientConfig() fakeosb.FakeClientConfiguration {
+	// up reactions for the fake components.
+	setup func(t *controllerTest)
+	// function to run just prior to creating the broker. This will only be run
+	// if there is actually a broker to create.
+	preCreateBroker func(t *controllerTest)
+	// function to run just after creating, and optionally verifying, the broker.
+	// This will only be run if there is actually a broker to create.
+	postCreateBroker func(t *controllerTest)
+	// function to run just prior to deleting the broker. This will only be run
+	// if there was a broker created.
+	preDeleteBroker func(t *controllerTest)
+	// function to run just after deleting the broker. This will only be run
+	// if there was a broker created.
+	postDeleteBroker func(t *controllerTest)
+	// function to run before creating the instance. This will only be run
+	// if there is actually an instance to create.
+	preCreateInstance func(t *controllerTest)
+	// function to run just after creating, and optionally verifying, the instance.
+	// This will only be run if there is actually an instance to create.
+	postCreateInstance func(t *controllerTest)
+	// function to run just prior to deleting the instance. This will only be run
+	// if there was an instance created.
+	preDeleteInstance func(t *controllerTest)
+	// function to run just after deleting the instance. This will only be run
+	// if there was an instance created.
+	postDeleteInstance func(t *controllerTest)
+	// function to run before creating the binding. This will only be run
+	// if there is actually a binding to create.
+	preCreateBinding func(t *controllerTest)
+	// function to run just after creating, and optionally verifying, the binding.
+	// This will only be run if there is actually a binding to create.
+	postCreateBinding func(t *controllerTest)
+	// function to run just prior to deleting the binding. This will only be run
+	// if there was a binding created.
+	preDeleteBinding func(t *controllerTest)
+	// function to run just after deleting the binding. This will only be run
+	// if there was a binding created.
+	postDeleteBinding func(t *controllerTest)
+}
+
+// run executes the test.
+//
+// Steps performed:
+// - create controller and API server with fakes
+// - call setup function
+// - create broker
+// - create instance
+// - create binding
+// - call supplied test function
+// - delete binding
+// - delete instance
+// - delete broker
+// - clean up controller and API server
+func (ct *controllerTest) run(test func(*controllerTest)) {
+	kubeClient, catalogClient, catalogClientConfig, osbClient, controller, informers, shutdownServer, shutdownController := newTestController(ct.t)
+	defer shutdownController()
+	defer shutdownServer()
+
+	ct.kubeClient = kubeClient
+	ct.catalogClient = catalogClient
+	ct.catalogClientConfig = catalogClientConfig
+	ct.osbClient = osbClient
+	ct.controller = controller
+	ct.informers = informers
+
+	ct.client = catalogClient.ServicecatalogV1beta1()
+
+	if ct.setup != nil {
+		ct.setup(ct)
```

Comment:
This setup function should be called within`newTestController` up above, before starting the controllers.

Reply:
Would you care to elaborate on why you think that the `setup` function needs to be run before `newTestController`? The types of things that are done in typical `setup` functions use objects that don't even exist until after `newTestController` is run. Is there something being done in the `setup` functions that you think is problematic to run after `newTestController`?

Reply:
It needs to be run inside, not before. It needs to run after some things, but before other things. After the controllers have been created, but before they have been started.

Reply:
Sorry. I was not clear on what elaboration I was looking for. I was hoping for some details on what the deficiencies are and not on what the proposed solution to the deficiencies are. I believe this is being discussed in other issues and in slack, so feel free to forego providing those details here.